File: frontend/lib/features/chat/message_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'message_model.freezed.dart';
part 'message_model.g.dart';

@freezed
class MessageModel with _$MessageModel {
  const factory MessageModel({
    required String id,
    required String chatType,
    required String chatId,
    required String content,
    @Default('text') String type,
    String? mediaUrl,
    required String senderId,
    required String senderName,
    String? senderAvatarUrl,
    @Default([]) List<ReactionModel> reactions,
    String? replyTo,
    @Default(false) bool isRead,
    required String createdAt,
    String? updatedAt,
  }) = _MessageModel;

  factory MessageModel.fromJson(Map<String, dynamic> json) =>
      _$MessageModelFromJson(json);

  factory MessageModel.fromApi(Map<String, dynamic> json) {
    final sender = json['sender'] as Map<String, dynamic>?;
    final replyTo = json['replyToId'] ?? json['replyTo'];
    return MessageModel.fromJson({
      ...json,
      'content': json['content'] ?? '',
      'type': json['messageType'] ?? json['type'] ?? 'text',
      'senderName': sender?['displayName'] ?? json['senderName'] ?? '',
      'senderAvatarUrl': sender?['avatarUrl'] ?? json['senderAvatarUrl'],
      'replyTo': replyTo is Map ? replyTo['id'] : replyTo,
      'reactions': json['reactions'] ?? const [],
    });
  }
}

@freezed
class ReactionModel with _$ReactionModel {
  const factory ReactionModel({
    required String emoji,
    required String userId,
    String? userName,
  }) = _ReactionModel;

  factory ReactionModel.fromJson(Map<String, dynamic> json) =>
      _$ReactionModelFromJson(json);
}

File: frontend/src/app/features/chat/models/message.model.ts
import { PublicUser } from '../../auth/models/auth.model';

export type ChatType = 'group' | 'channel';

export type MessageType = 'text' | 'image' | 'video' | 'file' | 'voice' | 'poll';

export interface MessageReaction {
  id: string;
  messageId: string;
  userId: string;
  emoji: string;
  createdAt: string;
}

export interface Message {
  id: string;
  chatType: ChatType;
  chatId: string;
  senderId: string;
  sender: PublicUser;
  content: string | null;
  messageType: MessageType;
  mediaUrl: string | null;
  mediaMetadata: Record<string, unknown> | null;
  replyToId: string | null;
  isEdited: boolean;
  isDeleted: boolean;
  reactions?: MessageReaction[];
  createdAt: string;
}

export type MessageStatus = 'sending' | 'failed';

export interface ChatMessage extends Message {
  status?: MessageStatus;
  localId?: string;
}

export interface SendMessagePayload {
  chatId: string;
  chatType: ChatType;
  content?: string;
  messageType?: MessageType;
  mediaUrl?: string;
  replyTo?: string;
}

export interface TypingEvent {
  userId: string;
  chatId: string;
}

export interface ReactionEvent {
  chatId: string;
  messageId: string;
  emoji: string;
  userId: string;
  reactions: MessageReaction[];
}

export interface ChatTarget {
  id: string;
  chatType: ChatType;
  name: string;
  avatarUrl: string | null;
  memberCount: number;
}
